File: bp-workout/Models/WorkoutProgramModels.swift
import Foundation

struct Exercise: Codable, Hashable, Identifiable {
    var id: String { name }
    let name: String
    let maxWeight: String
    /// Prescribed working sets for this lift; omit in JSON to default to 3.
    let targetSets: Int?
    /// Reps per working set when not AMRAP; omit to infer from load text / history (default 8 in prefill).
    let targetReps: Int?
    /// Exercises sharing the same positive group index are supersetted; omit or null if not in a superset.
    let supersetGroup: Int?
    /// When true, working sets are AMRAP (reps to failure); omit or false for a fixed rep target.
    let isAmrap: Bool?
    /// When true, this line is warm-up / activation only (optional for save warnings).
    let isWarmup: Bool?
    /// Optional free-form notes for this exercise (coaching cues, equipment, etc.).
    let notes: String?

    var prescriptionIsAmrap: Bool { isAmrap == true }
    var prescriptionIsWarmup: Bool { isWarmup == true }

    var trimmedProgramNotes: String? {
        guard let n = notes else { return nil }
        let t = n.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
    }

    /// Clamped prescription used for logging UI (1…20).
    var prescribedSets: Int {
        let n = targetSets ?? 3
        return max(1, min(n, 20))
    }

    /// Prescribed rep target for fixed-rep work; nil when AMRAP or not specified in the program.
    var prescribedRepTarget: Int? {
        guard !prescriptionIsAmrap else { return nil }
        guard let r = targetReps else { return nil }
        return max(1, min(r, 100))
    }
}

struct WorkoutDay: Codable, Hashable, Identifiable {
    var id: String { label }
    let label: String
    let exercises: [Exercise]
}

struct WorkoutProgram: Codable, Hashable, Identifiable {
    let id: String
    let name: String
    let subtitle: String
    let period: String
    let dateRange: String
    let days: [WorkoutDay]
    let color: String
    let isUserCreated: Bool?
    /// Populated for server catalog programs with a marketplace category.
    let categorySlug: String?
    let categoryTitle: String?

    /// Marketplace accent when the program editor does not expose a custom color.
    static let defaultAccentHex = "#66bfcc"
}

struct CatalogCategory: Codable, Hashable, Identifiable {
    var id: String { slug }
    let slug: String
    let title: String
    let subtitle: String
    let sortOrder: Int
    let iconSfSymbol: String
}

struct ProgramStats: Codable, Hashable {
    let totalPrograms: Int
    let totalMonths: Int
    let totalWorkoutDays: Int
    let dateRange: String
}

struct WorkoutProgramsBundle: Codable {
    let programs: [WorkoutProgram]
    let stats: ProgramStats
    let categories: [CatalogCategory]

    enum CodingKeys: String, CodingKey {
        case programs
        case stats
        case categories
    }

    init(programs: [WorkoutProgram], stats: ProgramStats, categories: [CatalogCategory]) {
        self.programs = programs
        self.stats = stats
        self.categories = categories
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        programs = try c.decode([WorkoutProgram].self, forKey: .programs)
        stats = try c.decode(ProgramStats.self, forKey: .stats)
        categories = try c.decodeIfPresent([CatalogCategory].self, forKey: .categories) ?? []
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(programs, forKey: .programs)
        try c.encode(stats, forKey: .stats)
        try c.encode(categories, forKey: .categories)
    }
}

File: bp_workout_flutter/lib/data/models/workout_program_models.dart
/// JSON shapes for `GET /v1/catalog/programs` (see `api/src/services/workout-catalog.ts`).

class Exercise {
  const Exercise({
    required this.name,
    required this.maxWeight,
    this.targetSets,
    this.targetReps,
    this.supersetGroup,
    this.isAmrap,
    this.isWarmup,
    this.notes,
  });

  final String name;
  final String maxWeight;
  final int? targetSets;
  final int? targetReps;
  final int? supersetGroup;
  final bool? isAmrap;
  final bool? isWarmup;
  final String? notes;

  factory Exercise.fromJson(Map<String, dynamic> j) => Exercise(
        name: j['name'] as String? ?? '',
        maxWeight: j['maxWeight'] as String? ?? '',
        targetSets: j['targetSets'] as int?,
        targetReps: j['targetReps'] as int?,
        supersetGroup: j['supersetGroup'] as int?,
        isAmrap: j['isAmrap'] as bool?,
        isWarmup: j['isWarmup'] as bool?,
        notes: j['notes'] as String?,
      );
}

class WorkoutDay {
  const WorkoutDay({required this.label, required this.exercises});

  final String label;
  final List<Exercise> exercises;

  factory WorkoutDay.fromJson(Map<String, dynamic> j) => WorkoutDay(
        label: j['label'] as String? ?? '',
        exercises: (j['exercises'] as List<dynamic>? ?? [])
            .map((e) => Exercise.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class WorkoutProgram {
  const WorkoutProgram({
    required this.id,
    required this.name,
    required this.subtitle,
    required this.period,
    required this.dateRange,
    required this.days,
    required this.color,
    this.isUserCreated,
    this.categorySlug,
    this.categoryTitle,
  });

  final String id;
  final String name;
  final String subtitle;
  final String period;
  final String dateRange;
  final List<WorkoutDay> days;
  final String color;
  final bool? isUserCreated;
  final String? categorySlug;
  final String? categoryTitle;

  factory WorkoutProgram.fromJson(Map<String, dynamic> j) => WorkoutProgram(
        id: j['id'] as String? ?? '',
        name: j['name'] as String? ?? '',
        subtitle: j['subtitle'] as String? ?? '',
        period: j['period'] as String? ?? '',
        dateRange: j['dateRange'] as String? ?? '',
        days: (j['days'] as List<dynamic>? ?? [])
            .map((e) => WorkoutDay.fromJson(e as Map<String, dynamic>))
            .toList(),
        color: j['color'] as String? ?? '#66bfcc',
        isUserCreated: j['isUserCreated'] as bool?,
        categorySlug: j['categorySlug'] as String?,
        categoryTitle: j['categoryTitle'] as String?,
      );
}

class CatalogCategory {
  const CatalogCategory({
    required this.slug,
    required this.title,
    required this.subtitle,
    required this.sortOrder,
    required this.iconSfSymbol,
  });

  final String slug;
  final String title;
  final String subtitle;
  final int sortOrder;
  final String iconSfSymbol;

  factory CatalogCategory.fromJson(Map<String, dynamic> j) => CatalogCategory(
        slug: j['slug'] as String? ?? '',
        title: j['title'] as String? ?? '',
        subtitle: j['subtitle'] as String? ?? '',
        sortOrder: j['sortOrder'] as int? ?? 0,
        iconSfSymbol: j['iconSfSymbol'] as String? ?? 'figure.strengthtraining.traditional',
      );
}

class ProgramStats {
  const ProgramStats({
    required this.totalPrograms,
    required this.totalMonths,
    required this.totalWorkoutDays,
    required this.dateRange,
  });

  final int totalPrograms;
  final int totalMonths;
  final int totalWorkoutDays;
  final String dateRange;

  factory ProgramStats.fromJson(Map<String, dynamic> j) => ProgramStats(
        totalPrograms: j['totalPrograms'] as int? ?? 0,
        totalMonths: j['totalMonths'] as int? ?? 0,
        totalWorkoutDays: j['totalWorkoutDays'] as int? ?? 0,
        dateRange: j['dateRange'] as String? ?? '',
      );
}

class WorkoutProgramsBundle {
  const WorkoutProgramsBundle({
    required this.programs,
    required this.stats,
    required this.categories,
  });

  final List<WorkoutProgram> programs;
  final ProgramStats stats;
  final List<CatalogCategory> categories;

  factory WorkoutProgramsBundle.fromJson(Map<String, dynamic> j) =>
      WorkoutProgramsBundle(
        programs: (j['programs'] as List<dynamic>? ?? [])
            .map((e) => WorkoutProgram.fromJson(e as Map<String, dynamic>))
            .toList(),
        stats: ProgramStats.fromJson(
          j['stats'] as Map<String, dynamic>? ?? {},
        ),
        categories: (j['categories'] as List<dynamic>? ?? [])
            .map((e) => CatalogCategory.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}
